packages/context/src/resolution-session.ts

```diff
@@ -0,0 +1,180 @@
+// Copyright IBM Corp. 2013, 2018. All Rights Reserved.
+// Node module: @loopback/context
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Binding} from './binding';
+import {Injection} from './inject';
+import * as debugModule from 'debug';
+import {DecoratorFactory} from '@loopback/metadata';
+
+const debugSession = debugModule('loopback:context:resolver:session');
+const getTargetName = DecoratorFactory.getTargetName;
+/**
+ * Object to keep states for a session to resolve bindings and their
+ * dependencies within a context
+ */
+export class ResolutionSession {
+  /**
+   * A stack of bindings for the current resolution session. It's used to track
+   * the path of dependency resolution and detect circular dependencies.
+   */
+  readonly bindings: Binding[] = [];
+
+  readonly injections: Injection[] = [];
+
+  /**
+   * Take a snapshot of the ResolutionSession so that we can pass it to
+   * `@inject.getter` without interferring with the current session
+   */
+  clone() {
+    const copy = new ResolutionSession();
+    copy.bindings.push(...this.bindings);
+    copy.injections.push(...this.injections);
+    return copy;
+  }
+
+  /**
+   * Start to resolve a binding within the session
+   * @param binding Binding
+   * @param session Resolution session
+   */
+  static enterBinding(
+    binding: Binding,
+    session?: ResolutionSession,
+  ): ResolutionSession {
+    session = session || new ResolutionSession();
+    session.enter(binding);
+    return session;
+  }
+
+  /**
+   * Push an injection into the session
+   * @param injection Injection
+   * @param session Resolution session
+   */
+  static enterInjection(
+    injection: Injection,
+    session?: ResolutionSession,
+  ): ResolutionSession {
+    session = session || new ResolutionSession();
+    session.enterInjection(injection);
+    return session;
+  }
+
+  static describeInjection(injection?: Injection) {
```

Comment:
should we add a return type here?

Reply:
TypeScript will infer the return type. But for clarity of APIs, we can make it explicit.